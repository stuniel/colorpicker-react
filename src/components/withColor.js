import React from 'react'
import { createColorObject } from '../utils/color'

export const ColorContext = React.createContext({
  color: createColorObject('#FFFFFF'),
  onChange: () => {}
})

export function withColor(Component) {
  return function ColoredComponent(props) {
    return (
      <ColorContext.Consumer>
        {({ color, onChange }) => <Component {...props} color={color} onChange={onChange} />}
      </ColorContext.Consumer>
    )
  }
}
