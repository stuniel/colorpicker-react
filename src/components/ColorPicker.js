import React from 'react'
import PropTypes from 'prop-types'

import { ColorContext } from './withColor'
import { createColorObject } from '../utils/color'

const propTypes = {
  children: PropTypes.node.isRequired,
  color: PropTypes.string,
  onChange: PropTypes.func
}

const defaultProps = {
  color: '#FFFFFF',
  onChange: () => {}
}

function ColorPicker(props) {
  const {
    children, color, onChange
  } = props

  return (
    <ColorContext.Provider
      value={{
        color: createColorObject(color),
        onChange
      }}
    >
      {children}
    </ColorContext.Provider>
  )
}

ColorPicker.propTypes = propTypes

ColorPicker.defaultProps = defaultProps

export default ColorPicker
