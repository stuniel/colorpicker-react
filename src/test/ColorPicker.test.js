import React from 'react'
import { mount, shallow } from 'enzyme'

import ColorPicker from '../components/ColorPicker'
import withColor from '../components/withColor'

const Box = withColor(props => <span {...props} />)

describe('<ColorPicker />', () => {
  it('renders children correctly', () => {
    const wrapper = shallow(<ColorPicker><span /></ColorPicker>)
    expect(wrapper).toMatchSnapshot()
  })
})

describe('color', () => {
  it('sets state.color correctly', () => {
    const color = '#FFFFFF'
    const wrapper = mount(<ColorPicker><Box /></ColorPicker>)
    expect(wrapper.state().color.hex.value).toEqual(color)
    wrapper.unmount()
  })

  it('sets state.color correctly when \'defaultColor\' is provided', () => {
    const color = '#123456'
    const wrapper = mount(<ColorPicker defaultColor={color}><Box /></ColorPicker>)
    expect(wrapper.state().color.hex.value).toEqual(color)
    wrapper.unmount()
  })

  it('passes \'#FFFFFF\' as \'defaultColor\' by default', () => {
    const color = '#FFFFFF'
    const wrapper = mount(<ColorPicker><Box /></ColorPicker>)
    const box = wrapper.find('span')
    expect(box.props().color.hex.value).toEqual(color)
  })

  it('passes \'defaultColor\' correctly', () => {
    const color = '#123456'
    const wrapper = mount(<ColorPicker defaultColor={color}><Box /></ColorPicker>)
    const box = wrapper.find('span')
    expect(box.props().color.hex.value).toEqual(color)
    wrapper.unmount()
  })

  it('updates \'defaultColor\' correctly', () => {
    const defaultColor = '#123456'
    const wrapper = mount(<ColorPicker><Box /></ColorPicker>)
    wrapper.setProps({ defaultColor })
    wrapper.update()
    const box = wrapper.find('span')
    expect(box.props().color.hex.value).toEqual(defaultColor)
    wrapper.unmount()
  })

  it('updates \'defaultColor\' correctly when it was previously provided', () => {
    const color = '#654321'
    const defaultColor = '#123456'
    const wrapper = mount(<ColorPicker defaultColor={color}><Box /></ColorPicker>)
    wrapper.setProps({ defaultColor })
    wrapper.update()
    const box = wrapper.find('span')
    expect(box.props().color.hex.value).toEqual(defaultColor)
    wrapper.unmount()
  })
})

describe('onChange', () => {
  it('passes \'onChange\' correctly', () => {
    const onChange = jest.fn()
    const wrapper = mount(<ColorPicker onChange={onChange}><Box /></ColorPicker>)
    const box = wrapper.find('span')
    box.simulate('change')
    expect(onChange).toHaveBeenCalledTimes(1)
    wrapper.unmount()
  })

  it('sets state.color correctly', () => {
    const color = { hex: { value: '#FF0000' } }
    const onChange = jest.fn().mockImplementation(n => n)
    const wrapper = mount(<ColorPicker onChange={onChange}><Box /></ColorPicker>)
    const box = wrapper.find('span')
    box.props().onChange(color)
    wrapper.update()
    expect(wrapper.state().color).toEqual(color)
    wrapper.unmount()
  })
})
