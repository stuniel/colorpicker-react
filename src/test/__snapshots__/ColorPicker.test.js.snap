// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ColorPicker /> renders children correctly 1`] = `
<[object Object]
  value={
    Object {
      "color": Object {
        "hex": Object {
          "value": "#FFFFFF",
        },
        "hsl": Object {
          "h": 0,
          "l": 1,
          "s": 0,
          "value": "hsl(0, 0%, 100%)",
        },
        "hsv": Object {
          "h": 0,
          "s": 0,
          "v": 1,
        },
        "rgb": Object {
          "b": 255,
          "g": 255,
          "r": 255,
          "value": "rgb(255, 255, 255)",
        },
      },
      "onChange": [Function],
    }
  }
>
  <span />
</[object Object]>
`;
